File: src/Programmers/Lv2/Q106/Q106_InterceptionSystem.java
/**
* Author    : Lee In Bok 
* Date      : 2023.11.17(Fri)
* Spend Time: overtime 30m
* Runtime   : 393.46 ms (Longest time)
* Memory    : 192 MB (Most spacious)
* Algoritm  : Implementation
 */

import java.util.Arrays;
import java.util.Comparator;

class Q106_InterceptionSystem{
    public static void main(String[] args) {
        Solution s = new Solution();
        int result = s.solution(new int[][]{{4,5},{4,8},{10,14},{11,13},{5,12},{3,7},{1,4}});
        
        System.out.println(result);
    }
}

class Solution {
    public int solution(int[][] targets) {
        int ans = 0;
        int bound = 0;

        Arrays.sort(targets, new Comparator<int[]>() {
            public int compare(int[] x, int[] y){
                return x[0] - y[0];
            }
        });

        for(int target[]: targets){
            if(bound > target[0]){
                bound = Math.min(bound, target[1]);
            } else {
                ans++;
                bound = target[1];
            }
        }

        return ans;
    }
}

File: CodeTest/src/Programmers/Lv2/Q106/Q106_InterceptionSystem.py
"""
# Author    : Lee In Bok 
# Date      : 2023.11.17(Fri)
# Spend Time: overtime 30m
# Runtime   : 1286.09 ms (Longest time)
# Memory    : 89.2 MB (Most spacious)
# Algoritm  : Implementation
"""

# def solution(targets):
#     ans, bound = 0, 0
#     targets = sorted(targets, key=lambda x: x[0])

#     for x, y in targets:
#         if bound > x:
#             bound = min(bound, y)
#         else:
#             bound = y
#             ans += 1

#     return ans

def solution(targets):
    ans, bound = 0, 0
    
    for x, y in sorted(targets):
        if x >= bound:
            bound = y
            ans += 1

    return ans

# def solution(targets):
#     ans, bound = 0, 0
    
#     for x, y in sorted(targets):
#         print(x, '    ', y, '    ', bound)
#         if x >= bound:
#             bound = y
#             ans += 1
#         else:
#             bound = min(y, bound)

#     return ans

result = solution([[0,4],[5,10],[6,8],[8,9]])

print(result)